language-support/ts/daml-ledger/index.ts

```diff
@@ -608,44 +628,147 @@ class Ledger {
-      // event when `contract` is currently not null. We should also only see
-      // archive events when `contract` is currently not null and the contract
-      // ids match. However, the JSON API does not provied these guarantees yet
-      // but we're working on them.
+    return this.streamFetchByKeyCommon(template, [key], "streamFetchByKey", (cs) => cs[0], (c) => [c]);
+  }
+
+  /**
+   * @internal
+   *
+   * Returns the same API as [[streamSubmit]] but does not, in fact, establish
+   * any socket connection. Instead, this is a stream that always has the given
+   * value as its accumulated state.
+   */
+  constantStream<T extends object, K, I extends string, V>(
+    value: V,
+  ): Stream<T, K, I, V> {
+    function on(t: 'live', l: (v: V) => void): void;
+    function on(t: 'change', l: (v: V, events: readonly Event<T, K, I>[]) => void): void;
+    function on(t: 'close', l: (e: StreamCloseEvent) => void): void;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    function on(type: any, listener: any): any {
+        if (type === 'live') {
+          listener(value);
+        }
+        if (type === 'change') {
+          listener(value, []);
+        }
+    }
+    function off(t: 'live', l: (v: V) => void): void;
+    function off(t: 'change', l: (v: V, events: readonly Event<T, K, I>[]) => void): void;
+    function off(t: 'close', l: (e: StreamCloseEvent) => void): void;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unused-vars, @typescript-eslint/no-empty-function
+    function off(_t: any, _l: any): any {}
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    return { on, off, close: (): void => {} };
+  }
+
+  /**
+   * @internal
+   *
+   */
+  streamFetchByKeyCommon<T extends object, K, I extends string, S>(
+    template: Template<T, K, I>,
+    keys: K[], // non-zero length
+    name: string,
+    to: (contracts: (CreateEvent<T, K, I> | null)[]) => S,
+    from: (state: S) => (CreateEvent<T, K, I> | null)[],
+  ): Stream<T, K, I, S> {
+    const lastContractIds: (ContractId<T> | null)[] = Array(keys.length).fill(null);
+    const keysCopy = Array.from(keys);
+    const initState: (CreateEvent<T, K, I> | null)[] = Array(keys.length).fill(null);
+    const request = keys.map(k => ({templateId: template.templateId, key: k}));
+    const reconnectRequest = (): object[] => request.map((r, idx) => ({...r, 'contractIdAtOffset': lastContractIds[idx]}))
+    const change = (state: S, events: readonly Event<T, K, I>[]): S => {
+      const contracts = from(state);
+      const newState: (CreateEvent<T, K, I> | null)[] = Array.from(contracts);
       for (const event of events) {
         if ('created' in event) {
-          contract = event.created;
+          const k = event.created.key;
+          keysCopy.forEach((requestKey, idx) => {
+            if (deepEqual(requestKey, k, {strict: true})) {
```

Comment:
Maybe worth factoring this out into a helper function that compares LF values for equality? It’s probably also worth pointing out that the JSON representation accepted by the JSON API is not unique (e.g. records can also be represented as an array). The reason this works anyway is that the typescript type system enforces that users use the same representation when creating a query that will be returned by the JSON API.

Reply:
I'm not quite sure what you want me to do here. It seems pointless to extract the method if it's just to write `keyEqual = (k1, k2) => deepEqual(k1, k2, {strict: true}` and then use it only once.

I could see a point in extracting it if we were using it in multiple places, or if we wanted to support more specifically LF equality across representations, but if the type system is enforcing a common one, what's the value in that?

Reply:
This was mostly just an idle thought. I briefly wondered if `deepEqual` was wrong and we need to worry about non-unique representation allowed by the JSON API. If this was just `lfValueEqual` or something like that, the reader doesn’t have to wonder if `deepEqual` is sufficient and we could add a comment to `lfValueEqual` (or whatever you want to call that) explaining why `deepEqual` is the right implementation.

Feel free to ignore my suggestion and follow your preference here.

Reply:
I do not know how to translate the JSON representation equivalence rules in a generic context. So I went for documenting that the given keys must be in "output" format instead, as that one is unique.

Reply:
Actually with the current implementation this is a breaking change for the `streamFetchByKey` function. So I'll revert that to its original form.

Reply:
You mean when called from JS where you can use a different format? I don’t think we should support a JS API that would produce type errors in typescript. It might work occasionally by accident but at least half of our code relies heavily on the assumption that you use the format dictated by the typescript code.

Reply:
I disagree. We advertize these bindings as JS bindings, not TS ones. We document the LF to JSON conversion as supporting multiple JSON formats for the same LF. Therefore existing JS code that uses a valid input encoding that differs from the output one is valid, and breaking that code is a breaking change.

Plus the cost of keeping that code working is really low here.

Reply:
Yes; this is why I firmly believe the right way to offer backwards compatibility is to not change existing code.